### DISCUSSION TITLE
Outputting WandB performance metrics

### DISCUSSION THREAD

**Discussion by canihaveabravo** [kitzeslab/opensoundscape#903] [General]
Hello, 

I'm having ongoing difficulty outputting wandb logs into useful performance metrics. Can you provide any tips for this? The wandb_session argument from the CNN training tutorial doesnt seem to run with my code. 

Thanks very much, 
Lorenzo

**Comment by sammlapp** [comment]
Hi Lorenzo, please provide some details of your code and error messages so that we can assist you. 

An example of using wandb with CNN.train is given here: https://opensoundscape.org/en/latest/tutorials/train_cnn.html#Set-up-WandB-model-logging 

Make sure you have created a wandb account and run `wandb.login()` with your API key, or `wandb login` in terminal, using the API key from https://wandb.ai/authorize 

It will also be worth checking if wandb logging works when you use the tutorial notebook from that link

**Reply by canihaveabravo** [reply]
Hi Samm, 

Yes I am able to log a run in WandB using the code below and I can see basic hyperparameter plots, but none of the interesting graphs of performance metrics, or samples described in the tutorial. Any ideas? 

import` wandb
#Import WandB Utility
try:
    wandb.login()
    wandb_session = wandb.init(
        entity='seiche',
        project='multispecies_CNN',
        name='multispecies_CNN',
    )

except: #if wandb.init fails, don't use wandb logging
    print('failed to create wandb session. wandb session will be None')
    wandb_session = `None'



However when inserting the wandb_session argument into my training loop:

model.train(
    balanced_train_df,
    valid_df,
    epochs= 10,
    batch_size= 128,
    log_interval = 5,
    num_workers= 0, # if multiple processors are available utilise them to speed up training
    wandb_session = wandb_session,
    save_interval= 10,
    save_path= checkpoint_folder #location to save checkpoints
)



I get the error:

**[c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\opensoundscape\audio.py:340](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/opensoundscape/audio.py:340): UserWarning: Audio object is shorter than requested duration: 10.000020833333334 sec instead of 30 sec
  warnings.warn(error_msg)
[c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\opensoundscape\audio.py:340](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/opensoundscape/audio.py:340): UserWarning: Audio object is shorter than requested duration: 12.0 sec instead of 30 sec
  warnings.warn(error_msg)
[c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\opensoundscape\audio.py:340](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/opensoundscape/audio.py:340): UserWarning: Audio object is shorter than requested duration: 10.000020833333334 sec instead of 30 sec
  warnings.warn(error_msg)
<frozen importlib._bootstrap>:1047: ImportWarning: ImportHookFinder.find_spec() not found; falling back to find_module()**
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
[c:\Users\lscala\Python_learning\multispecies_trial_chris.ipynb](file:///C:/Users/lscala/Python_learning/multispecies_trial_chris.ipynb) Cell 20 line 1
     [12](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=11) model.optimizer_params['weight_decay']=0.0005
     [14](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=13) # Train the model on the full training set and validate on the validation set. Save the model to directory.
---> [15](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=14) model.train(
     [16](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=15)     balanced_train_df,
     [17](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=16)     valid_df,
     [18](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=17)     epochs= 2,
     [19](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=18)     batch_size= 128,
     [20](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=19)     log_interval = 5,
     [21](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=20)     num_workers= 0, # if multiple processors are available utilise them to speed up training
     [22](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=21)     wandb_session = wandb_session,
     [23](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=22)     save_interval= 10,
     [24](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=23)     save_path= checkpoint_folder #location to save checkpoints
     [25](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=24) )
     [27](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=26) #Let Wandb know that we finished training successfully
     [28](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=27) wandb.unwatch(model.network)

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\opensoundscape\ml\cnn.py:841](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/opensoundscape/ml/cnn.py:841), in CNN.train(self, train_df, validation_df, epochs, batch_size, num_workers, save_path, save_interval, log_interval, validation_interval, invalid_samples_log, raise_errors, wandb_session, progress_bar)
    833         wandb_session.watch(
    834             self.network,
    835             log="all",
    836             log_freq=log_freq,
    837             log_graph=(self.wandb_logging["log_graph"]),
...
-> 1617     with open(path, mode, encoding=encoding) as f:
   1618         yield f
   1620         f.flush()

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\lscala\\AppData\\Local\\Temp\\tmplk94vkyk\\Samples / training samples.table.json'

**Reply by sammlapp** [reply]
interesting, this seems like a wandb error. Your error message is not showing fully (the relevant part is within the `...`) so it's hard to debug, but it seems like a caching error from wandb. It's possible that deleting the temporary folder `C:\Users\lscala\AppData\Local\Temp\tmplk94vkyk` could help.

**Reply by canihaveabravo** [reply]
Deleting the temp folder didnt help, but here is the error message in full. Thank you very much for your help.

---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
[c:\Users\lscala\Python_learning\multispecies_trial_chris.ipynb](file:///C:/Users/lscala/Python_learning/multispecies_trial_chris.ipynb) Cell 20 line 1
     [12](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=11) model.optimizer_params['weight_decay']=0.0005
     [14](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=13) # Train the model on the full training set and validate on the validation set. Save the model to directory.
---> [15](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=14) model.train(
     [16](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=15)     balanced_train_df,
     [17](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=16)     valid_df,
     [18](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=17)     epochs= 2,
     [19](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=18)     batch_size= 128,
     [20](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=19)     log_interval = 5,
     [21](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=20)     num_workers= 0, # if multiple processors are available utilise them to speed up training
     [22](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=21)     wandb_session = wandb_session,
     [23](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=22)     save_interval= 10,
     [24](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=23)     save_path= checkpoint_folder #location to save checkpoints
     [25](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=24) )
     [27](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=26) #Let Wandb know that we finished training successfully
     [28](vscode-notebook-cell:/c%3A/Users/lscala/Python_learning/multispecies_trial_chris.ipynb#X25sZmlsZQ%3D%3D?line=27) wandb.unwatch(model.network)

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\opensoundscape\ml\cnn.py:841](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/opensoundscape/ml/cnn.py:841), in CNN.train(self, train_df, validation_df, epochs, batch_size, num_workers, save_path, save_interval, log_interval, validation_interval, invalid_samples_log, raise_errors, wandb_session, progress_bar)
    833         wandb_session.watch(
    834             self.network,
    835             log="all",
    836             log_freq=log_freq,
    837             log_graph=(self.wandb_logging["log_graph"]),
    838         )
    840     # log tables of preprocessed samples
--> 841     wandb_session.log(
    842         {
    843             "Samples / training samples": wandb_table(
    844                 AudioFileDataset(
    845                     train_df, self.preprocessor, bypass_augmentations=False
    846                 ),
    847                 self.wandb_logging["n_preview_samples"],
    848             ),
    849             "Samples / training samples no aug": wandb_table(
    850                 AudioFileDataset(
    851                     train_df, self.preprocessor, bypass_augmentations=True
    852                 ),
    853                 self.wandb_logging["n_preview_samples"],
    854             ),
    855             "Samples / validation samples": wandb_table(
    856                 AudioFileDataset(
    857                     validation_df, self.preprocessor, bypass_augmentations=True
    858                 ),
    859                 self.wandb_logging["n_preview_samples"],
    860             ),
    861         }
    862     )
    864 # Move network to device
    865 self.network.to(self.device)

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\wandb\sdk\wandb_run.py:368](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_run.py:368), in _run_decorator._noop.<locals>.wrapper(self, *args, **kwargs)
    365         wandb.termwarn(message, repeat=False)
    366         return cls.Dummy()
--> 368 return func(self, *args, **kwargs)

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\wandb\sdk\wandb_run.py:331](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_run.py:331), in _run_decorator._attach.<locals>.wrapper(self, *args, **kwargs)
    329         raise e
    330     cls._is_attaching = ""
--> 331 return func(self, *args, **kwargs)

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\wandb\sdk\wandb_run.py:1712](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_run.py:1712), in Run.log(self, data, step, commit, sync)
   1705 if sync is not None:
   1706     deprecate.deprecate(
   1707         field_name=deprecate.Deprecated.run__log_sync,
   1708         warning_message=(
   1709             "`sync` argument is deprecated and does not affect the behaviour of `wandb.log`"
   1710         ),
   1711     )
-> 1712 self._log(data=data, step=step, commit=commit)

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\wandb\sdk\wandb_run.py:1494](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_run.py:1494), in Run._log(self, data, step, commit)
   1491 if any(not isinstance(key, str) for key in data.keys()):
   1492     raise ValueError("Key values passed to `wandb.log` must be strings.")
-> 1494 self._partial_history_callback(data, step, commit)
   1496 if step is not None:
   1497     if os.getpid() != self._init_pid or self._is_attached:

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\wandb\sdk\wandb_run.py:1364](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_run.py:1364), in Run._partial_history_callback(self, row, step, commit)
   1361 if self._backend and self._backend.interface:
   1362     not_using_tensorboard = len(wandb.patched["tensorboard"]) == 0
-> 1364     self._backend.interface.publish_partial_history(
   1365         row,
   1366         user_step=self._step,
   1367         step=step,
   1368         flush=commit,
   1369         publish_step=not_using_tensorboard,
   1370     )

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\wandb\sdk\interface\interface.py:567](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/interface/interface.py:567), in InterfaceBase.publish_partial_history(self, data, user_step, step, flush, publish_step, run)
    556 def publish_partial_history(
    557     self,
    558     data: dict,
   (...)
    563     run: Optional["Run"] = None,
    564 ) -> None:
    565     run = run or self._run
--> 567     data = history_dict_to_json(run, data, step=user_step, ignore_copy_err=True)
    568     data.pop("_step", None)
    570     # add timestamp to the history request, if not already present
    571     # the timestamp might come from the tensorboard log logic

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\wandb\sdk\data_types\utils.py:52](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/data_types/utils.py:52), in history_dict_to_json(run, payload, step, ignore_copy_err)
     48         payload[key] = history_dict_to_json(
     49             run, val, step=step, ignore_copy_err=ignore_copy_err
     50         )
     51     else:
---> 52         payload[key] = val_to_json(
     53             run, key, val, namespace=step, ignore_copy_err=ignore_copy_err
     54         )
     56 return payload

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\wandb\sdk\data_types\utils.py:152](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/data_types/utils.py:152), in val_to_json(run, key, val, namespace, ignore_copy_err)
    148     sanitized_key = re.sub(r"[^a-zA-Z0-9_]+", "", key)
    149     art = wandb.wandb_sdk.wandb_artifacts.Artifact(
    150         f"run-{run.id}-{sanitized_key}", "run_table"
    151     )
--> 152     art.add(val, key)
    153     run.log_artifact(art)
    155 # Partitioned tables and joined tables do not support being bound to runs.

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\wandb\sdk\wandb_artifacts.py:557](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_artifacts.py:557), in Artifact.add(self, obj, name)
    555         do_write(tmp_f)
    556 else:
--> 557     with self.new_file(name) as f:
    558         file_path = f.name
    559         do_write(f)

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\contextlib.py:137](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/contextlib.py:137), in _GeneratorContextManager.__enter__(self)
    135 del self.args, self.kwds, self.func
    136 try:
--> 137     return next(self.gen)
    138 except StopIteration:
    139     raise RuntimeError("generator didn't yield") from None

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\wandb\sdk\wandb_artifacts.py:390](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_artifacts.py:390), in Artifact.new_file(self, name, mode, encoding)
    388 filesystem.mkdir_exists_ok(os.path.dirname(path))
    389 try:
--> 390     with util.fsync_open(path, mode, encoding) as f:
    391         yield f
    392 except UnicodeEncodeError as e:

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\contextlib.py:137](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/contextlib.py:137), in _GeneratorContextManager.__enter__(self)
    135 del self.args, self.kwds, self.func
    136 try:
--> 137     return next(self.gen)
    138 except StopIteration:
    139     raise RuntimeError("generator didn't yield") from None

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\wandb\util.py:1617](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/util.py:1617), in fsync_open(path, mode, encoding)
   1612 @contextlib.contextmanager
   1613 def fsync_open(
   1614     path: Union[pathlib.Path, str], mode: str = "w", encoding: Optional[str] = None
   1615 ) -> Generator[IO[Any], None, None]:
   1616     """Open a path for I/O and guarante that the file is flushed and synced."""
-> 1617     with open(path, mode, encoding=encoding) as f:
   1618         yield f
   1620         f.flush()

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\lscala\\AppData\\Local\\Temp\\tmpqt2c4dg7\\Samples / training samples.table.json'

**Reply by sammlapp** [reply]
It seems like other wandb users have experiences a similar issue: https://github.com/wandb/wandb/issues/1869 (even since that issue was "closed") 

Are you perhaps able to run wandb.finish() then start a new wandb session with wandb.init() ? Does the error occur again?

**Comment by canihaveabravo** [comment]
Yeah unfortunately it occurs every time I start a new wandb_session.


Lorenzo Scala
Senior Marine Scientist
Seiche Ltd
T:
 +44(0)1409 404050
E: ***@***.***
W: https://www.seiche.com
Bradworthy Industrial Estate, Langdon Road, Bradworthy, Holsworthy, Devon, EX22 7SF, United Kingdom
 Seiche Ltd is registered in England & Wales. Registered number: 1468514. Registered Office: Bradworthy Industrial Estate, Langdon Road, Bradworthy, Holsworthy, Devon, EX22 7SF, United Kingdom
 
 The information contained in this e-mail transmission, and any documents, files or previous e-mail messages attached to it, is privileged and confidential, and solely intended for the use of the individual(s) to whom it is addressed.  If you are not the intended recipient or a person responsible for delivering it to the intended recipient you should not read, copy, distribute or otherwise use the information, and you are hereby notified that any disclosure, copying, distribution or use of any of the information contained in or attached to this message is strictly prohibited.  If you have received this message in error, please notify sender immediately and delete this e-mail and attached documents.

**Reply by sammlapp** [reply]
can you make sure you have the latest version of `wandb` installed in your environment?

**Comment by canihaveabravo** [comment]
I have updated to wandb version 0.16.0 but the issue persists. I am scratching my head a bit with it 😊


Lorenzo Scala
Senior Marine Scientist
Seiche Ltd
T:
 +44(0)1409 404050
E: ***@***.***
W: https://www.seiche.com
Bradworthy Industrial Estate, Langdon Road, Bradworthy, Holsworthy, Devon, EX22 7SF, United Kingdom
 Seiche Ltd is registered in England & Wales. Registered number: 1468514. Registered Office: Bradworthy Industrial Estate, Langdon Road, Bradworthy, Holsworthy, Devon, EX22 7SF, United Kingdom
 
 The information contained in this e-mail transmission, and any documents, files or previous e-mail messages attached to it, is privileged and confidential, and solely intended for the use of the individual(s) to whom it is addressed.  If you are not the intended recipient or a person responsible for delivering it to the intended recipient you should not read, copy, distribute or otherwise use the information, and you are hereby notified that any disclosure, copying, distribution or use of any of the information contained in or attached to this message is strictly prohibited.  If you have received this message in error, please notify sender immediately and delete this e-mail and attached documents.

**Reply by sammlapp** [reply]
Sorry, I also don't know what is causing the issue and it seems to be caused by wandb rather than opensoundscape. I suspect some sort of "cache clearing" like deleting temporary or cached files will resolve the issue

**Reply by sammlapp** [reply]
@canihaveabravo is it possible your disk is nearly full? See this thread: https://github.com/wandb/wandb/issues/1869#issuecomment-1801177268

**Reply by sammlapp** [reply]
@canihaveabravo I encountered a similar issue recently. I resolved it by deleting all folders in the `temp` folder that had `wandb` in the name and killing all `wandb` processes (can check for them using `htop` or a similar tool).

**Comment by canihaveabravo** [comment]
Hi Sam,

Thanks for the tip. I’ve tried this but unfortunately no luck so far.


Lorenzo Scala
Senior Marine Scientist
Seiche Ltd
T:
 +44(0)1409 404050
E: ***@***.***
W: https://www.seiche.com
Bradworthy Industrial Estate, Langdon Road, Bradworthy, Holsworthy, Devon, EX22 7SF, United Kingdom
 Seiche Ltd is registered in England & Wales. Registered number: 1468514. Registered Office: Bradworthy Industrial Estate, Langdon Road, Bradworthy, Holsworthy, Devon, EX22 7SF, United Kingdom
 
 The information contained in this e-mail transmission, and any documents, files or previous e-mail messages attached to it, is privileged and confidential, and solely intended for the use of the individual(s) to whom it is addressed.  If you are not the intended recipient or a person responsible for delivering it to the intended recipient you should not read, copy, distribute or otherwise use the information, and you are hereby notified that any disclosure, copying, distribution or use of any of the information contained in or attached to this message is strictly prohibited.  If you have received this message in error, please notify sender immediately and delete this e-mail and attached documents.

**Reply by sammlapp** [reply]
If you still need help resolving this to move forward, send me an email and we can try to tackle it together

**Comment by canihaveabravo** [comment]
Hi Sam. Yes I have no yet found a solution to this and have therefore taken another direction. However, happy to return to it and tackle it together.

**Comment by fascimare** [comment]
If you found a solution, please let me know. I am having the same issue.

**Reply by sammlapp** [reply]
Hi @fascimare sorry for not getting back to you. Can you also share your full error text to assist with debugging?

**Reply by sammlapp** [reply]
Also, it would be helpful if you can update wandb to the newest version and check if the error still occurs. 
```
conda activate environment_name
pip install --upgrade wandb
```

**Comment by sammlapp** [comment]
@canihaveabravo and @fascimare , wandb seems to have several issues that could be causing this, but its worth checking if updating your wandb package to the most recent version fixes it. I still haven't been able to reproduce the behavior myself, unfortunately, so its difficult to debug. If you do try with the newest wandb version, I would love to hear if it works or not (please also post the version you install)

**Comment by m1alksne** [comment]
Hi All, 

I have previously trained a model with opensoundscape and logged the metrics in real time with wandb. However, when I did that, I was using a Linux machine. I am currently trying to reproduce this work using a Windows machine. Originally when I tried this, I got an error messge similar to @canihaveabravo. When I updated my wandb package and made sure to activate my conda environment, I no longer got the error message.  

I simply typed:

conda activate environment_name
pip install --upgrade wandb
print(wandb.__version__)
0.16.2

Unfortunately, although my code is no longer throwing an error message caused by wandb not finding the correct temp file, I am still unable to visualize the metrics using the wandb interface.  The run does not crash but no metrics get logged or plotted. So, I didn't solve the problem but maybe got a little bit closer. I hope this helps.

**Comment by sammlapp** [comment]
it would also be helpful to know what operating system (and version) everyone is on @m1alksne @canihaveabravo @fascimare 

our current hypothesis is that this is a Windows-specific problem

**Reply by sammlapp** [reply]
Possibly related issues on wandb:
https://github.com/wandb/wandb/issues/6896
https://github.com/wandb/wandb/issues/6704
https://github.com/wandb/wandb/issues/6784

in particular, it might be worth trying to turn off logging of gradients (`my_cnn.wandb_logging['log_graph']=False`)

**Reply by m1alksne** [reply]
Hi Sam, I am using a Windows10 machine.

**Reply by canihaveabravo** [reply]
my_cnn.wandb_logging['log_graph']=False

this appears to help a bit. I dont get the same error message anymore, however I do get the below:

AttributeError                            Traceback (most recent call last)
Cell In[23], [line 15](vscode-notebook-cell:?execution_count=23&line=15)
      [2](vscode-notebook-cell:?execution_count=23&line=2) model.train(
      [3](vscode-notebook-cell:?execution_count=23&line=3)     balanced_train_df,
      [4](vscode-notebook-cell:?execution_count=23&line=4)     valid_df,
   (...)
     [11](vscode-notebook-cell:?execution_count=23&line=11)     save_path= checkpoint_folder #location to save checkpoints
     [12](vscode-notebook-cell:?execution_count=23&line=12) )
     [14](vscode-notebook-cell:?execution_count=23&line=14) #Let Wandb know that we finished training successfully
---> [15](vscode-notebook-cell:?execution_count=23&line=15) wandb.unwatch(model.network)
     [16](vscode-notebook-cell:?execution_count=23&line=16) wandb.finish()

File [c:\Users\lscala\AppData\Local\anaconda3\envs\opensoundscape\Lib\site-packages\wandb\sdk\wandb_watch.py:123](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_watch.py:123), in unwatch(models)
    [121](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_watch.py:121)         else:
    [122](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_watch.py:122)             for name in model._wandb_hook_names:
--> [123](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_watch.py:123)                 wandb.run._torch.unhook(name)
    [124](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_watch.py:124)             delattr(model, "_wandb_hook_names")
    [125](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_watch.py:125)             # TODO: we should also remove recursively model._wandb_watch_called
    [126](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_watch.py:126) 
    [127](file:///C:/Users/lscala/AppData/Local/anaconda3/envs/opensoundscape/Lib/site-packages/wandb/sdk/wandb_watch.py:127) else:

AttributeError: 'NoneType' object has no attribute '_torch'

**Reply by fascimare** [reply]
Apologies for the late reply. I'm on a Windows 11 computer. I upgraded wandb to version 0.16.3 and turned off the logging of gradients, but still get the following error message:
Traceback (most recent call last):

  Cell In[14], line 1
    model.train(

  File ~\AppData\Local\anaconda3\Lib\site-packages\opensoundscape\ml\cnn.py:841 in train
    wandb_session.log(

  File ~\AppData\Local\anaconda3\Lib\site-packages\wandb\sdk\wandb_run.py:420 in wrapper
    return func(self, *args, **kwargs)

  File ~\AppData\Local\anaconda3\Lib\site-packages\wandb\sdk\wandb_run.py:371 in wrapper_fn
    return func(self, *args, **kwargs)

  File ~\AppData\Local\anaconda3\Lib\site-packages\wandb\sdk\wandb_run.py:361 in wrapper
    return func(self, *args, **kwargs)

  File ~\AppData\Local\anaconda3\Lib\site-packages\wandb\sdk\wandb_run.py:1828 in log
    self._log(data=data, step=step, commit=commit)

  File ~\AppData\Local\anaconda3\Lib\site-packages\wandb\sdk\wandb_run.py:1596 in _log
    self._partial_history_callback(data, step, commit)

  File ~\AppData\Local\anaconda3\Lib\site-packages\wandb\sdk\wandb_run.py:1468 in _partial_history_callback
    self._backend.interface.publish_partial_history(

  File ~\AppData\Local\anaconda3\Lib\site-packages\wandb\sdk\interface\interface.py:558 in publish_partial_history
    data = history_dict_to_json(run, data, step=user_step, ignore_copy_err=True)

  File ~\AppData\Local\anaconda3\Lib\site-packages\wandb\sdk\data_types\utils.py:52 in history_dict_to_json
    payload[key] = val_to_json(

  File ~\AppData\Local\anaconda3\Lib\site-packages\wandb\sdk\data_types\utils.py:156 in val_to_json
    art.add(val, key)

  File ~\AppData\Local\anaconda3\Lib\site-packages\wandb\sdk\artifacts\artifact.py:1375 in add
    with self.new_file(name) as f:

  File ~\AppData\Local\anaconda3\Lib\contextlib.py:137 in __enter__
    return next(self.gen)

  File ~\AppData\Local\anaconda3\Lib\site-packages\wandb\sdk\artifacts\artifact.py:1122 in new_file
    with util.fsync_open(path, mode, encoding) as f:

  File ~\AppData\Local\anaconda3\Lib\contextlib.py:137 in __enter__
    return next(self.gen)

  File ~\AppData\Local\anaconda3\Lib\site-packages\wandb\util.py:1513 in fsync_open
    with open(path, mode, encoding=encoding) as f:

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\P310512\\AppData\\Local\\Temp\\tmpyqt9fy2m\\Samples / training samples.table.json'

**Reply by fascimare** [reply]
I also tried the debug of the first blogpost you referred to, but no luck.

**Reply by sammlapp** [reply]
Ok thanks for checking. I'm curious, do you have a folder "C:\Users\P310512\AppData\Local\Temp\tmpyqt9fy2m" and if so what files are in it?

**Reply by fascimare** [reply]
The folder is there, and in it a folder 'Samples', but no other files and the Samples folder is empty.

**Comment by sammlapp** [comment]
Following @louisfh 's suggestion, I created a branch of opensoundscape where some special characters are removed from wandb logging calls (especially the "/" character). The branch is called `patch_wandb_windows`. If any of you @canihaveabravo  @fascimare @m1alksne have a chance to try it out, I'm curious if it will resolve the wandb logging issue that we haven't been able to reproduce on our end. 

To try it, you'll need to clone the opensoundscape github, checkout the branch, and pip install it. Here is an example of command line operations:
```
cd
git clone git@github.com:kitzeslab/opensoundscape.git
cd opensoundscape
git checkout patch_wandb_windows
git pull
```
now create a new conda environment (assuming you use conda for environment management)
```
conda deactivate
conda create -n opso_env_with_patch python=3.10
```
you might have to enter 'y' here to confirm you want to create the new conda environment
conda activate opso_env_with_patch. 

You can install the package of the directory you're currently in with `pip install .` after activating the new environment
```
conda activate opso_env_with_patch
pip install .
```

Then check if you get the same errors when using this environment. 

Fingers crossed that this helps!

**Reply by fascimare** [reply]
I tried it out today, and it's working! The only thing I'm not getting are the preview samples, but the rest is there. Thank you so much @sammlapp !

**Reply by m1alksne** [reply]
Hey @fascimare what type of Windows machine are you running on? I have not been able to get the wandb plots to work. Although I followed these steps and the code no longer throws an error.
